cli/components/list_raw_object.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package components
+
+import (
+	"fmt"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/rs/zerolog/log"
+	"go.mondoo.com/cnquery/v11/cli/theme"
+)
+
+// Object is the interface that a list need to implement so we can display its objects.
+type Object interface {
+	// PrintableKeys returns the list of keys that will be printed.
+	PrintableKeys() []string
+
+	// PrintableValue returns the key value based of the provided index.
+	PrintableValue(index int) string
+}
+
+// List is a non-interactive function that lists objects to the user.
+//
+// e.g.
+// ```go
+//
+//	type CustomString string
+//
+//	func (s CustomString) PrintableKeys() []string {
+//		return []string{"string"}
+//	}
+//	func (s CustomString) PrintableValue(_ int) string {
+//		return string(s)
+//	}
+//
+//	func main() {
+//		customStrings := []CustomString{"first", "second", "third"}
+//		list := components.List(theme.OperatingSystemTheme, "string", customStrings)
+//		fmt.Printf(list)
+//	}
+//
+// ```
+func List[O Object](theme *theme.Theme, listType string, list []O) string {
+	b := &strings.Builder{}
+	w := tabwriter.NewWriter(b, 1, 1, 1, ' ', tabwriter.TabIndent)
+
+	log.Debug().Msgf("discovered %d %s(s)", len(list), listType)
+
+	for i := range list {
+		assetObj := list[i]
+
+		for i, key := range assetObj.PrintableKeys() {
```

Comment:
I am not sure I am in favour of this construction. We are building a generic way of displaying lists of items and selecting an item from that list. I think the way an item from the list is displayed, should be defined outside of this function. The current implementation seems like it provides some "customization" options but it is actually very limited.

If you don't need the keys and values individually, why not just make the interface for `Object` have a function `Print() string` or something similar, and just call this here. Then it is up to the `Object` to specify how it is printed on the CLI. We also remove the `PrintableKeys` and `PrintableValue`.

Reply:
I disagree.

Leaving the formatting to the implementation (caller) is not an option, if we want that, we should not have this function at all.

The whole point of these helpers is to help and standardize how we print/interact with "stuff".